test/e2e/storage/testsuites/volumeperf.go

```diff
@@ -0,0 +1,287 @@
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epv "k8s.io/kubernetes/test/e2e/framework/pv"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+)
+
+type volumePerformanceTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+var _ TestSuite = &volumePerformanceTestSuite{}
+
+const (
+	provisionCount     = 500
+	testTimeout        = 15 * time.Minute
+	expectedLatency    = 5 * time.Minute
+	expectedThroughput = float64(1)
+)
+
+// Time intervals when a resource was created, it enters
+// the desired state and the elapsed time between these
+// two states
+type interval struct {
+	create            time.Time
+	enterDesiredState time.Time
+	elapsed           time.Duration
+}
+
+// Metrics to evaluate performance of an operation
+// TODO: Add metrics like median, mode, standard deviation, percentile
+type metrics struct {
+	latency    time.Duration
+	throughput float64
+}
+
+// Consolidates performance stats for any operation
+type performanceStats struct {
+	mutex             *sync.Mutex
+	perObjectInterval map[string]*interval
+	operationMetrics  *metrics
+}
+
+// waitForProvisionCh receives a signal from controller
+// when all PVCs are Bound
+// The test blocks until waitForProvisionCh receives a signal
+// or the test times out
+var waitForProvisionCh chan struct{}
+
+// InitVolumePerformanceTestSuite returns volumePerformanceTestSuite that implements TestSuite interface
+func InitVolumePerformanceTestSuite() TestSuite {
+	return &volumePerformanceTestSuite{
+		tsInfo: TestSuiteInfo{
+			Name: "volume-lifecycle-performance",
+			TestPatterns: []testpatterns.TestPattern{
+				testpatterns.FsVolModeDynamicPV,
+			},
+		},
+	}
+}
+
+func (t *volumePerformanceTestSuite) GetTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *volumePerformanceTestSuite) SkipRedundantSuite(driver TestDriver, pattern testpatterns.TestPattern) {
+}
+
+func (t *volumePerformanceTestSuite) DefineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config      *PerTestConfig
+		testCleanup func()
+		cs          clientset.Interface
+		ns          *v1.Namespace
+		pvcs        []*v1.PersistentVolumeClaim
+	}
+	ginkgo.BeforeEach(func() {
+		// Check preconditions
+		if dDriver = driver.(DynamicPVTestDriver); dDriver == nil {
+			framework.Failf("Test driver does not support dynamically created volumes")
+		}
+	})
+
+	f := framework.NewDefaultFramework("volume-lifecycle-performance")
+	ginkgo.It("volume lifecycle performance [Slow]", func() {
+		l := &local{
+			cs: f.ClientSet,
+			ns: f.Namespace,
+		}
+		l.config, l.testCleanup = driver.PrepareTest(f)
+		defer l.testCleanup()
+
+		// Stats for volume provisioning operation
+		// TODO: Add stats for attach, resize and snapshot?
```

Comment:
+1

Reply:
@xing-yang thinking of adding this in a follow up change to keep the PR size manageable..

Reply:
Yeah, that's reasonable.